### PR TITLE
BIG Refactor of the whole project.

### PR DESCRIPTION
**Pull request by ScottCTD**
- Add a resource manager to manage our resources.
- Delete City and all its stuff.
- Redo layout.
- Cleanup code bases.
- Add many docstrings.
- Optimize something.
- Update datasets.
- Redesign program running order.

**But, there are still many aspects that need to be improved.**

I will list them later maybe in the project, and we will continue working on it.

### CODE REVIEW

File: Application/resource_manager.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    if resource.is_complete():
+        resource.is_init = True
+        return True
+    else:
+        for _ in range(RETRY_COUNT):
+            resource.download()
+            resource.generate_identifier()
+            if resource.is_complete():
+                resource.is_init = True
+                return True
```
Comment (Lei-Tin) [inline Application/resource_manager.py:178]: Maybe log the 3 different tries? So that the user know what's going on

**Review by Lei-Tin — APPROVED** [review]
最大的那个 GUI Refactor 我不太确定改了什么

如果只是复制粘贴原来的然后 分成 Utils and Main then Great!
其他部分我看上去没发现什么问题

只是那个 COVID19 Global csv 删了 281 行 加了 281 行我不知道是 GitHub 有问题 or smth. 应该没事

File: Application/data.py [hunk 2]
```diff
@@ -1,21 +1,17 @@
...
-
-Contains functions to do all processings with our datasets
-Also contains CONSTANTS and classes
+This module contains all classes, functions, and constants for our data.
 """
+# Python built-ins
 import csv
 import datetime
-import hashlib
+import logging
```
[the remark this answers is not in the document]
Comment (Lei-Tin) [reply Application/data.py:7]: Plus 你在这边直接 logging.info 输出，main 那边的 logger 能接受到吗

File: Application/data.py [hunk 3]
```diff
@@ -239,18 +217,20 @@ def __str__(self) -> str:
 # All countries from our datasets.
 COUNTRIES: Set[Country] = set()
 SORTED_COUNTRIES: List[Country] = []
+KEY_COUNTRIES: List[Country] = [
```
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Application/data.py:220]: yeh but plus Canada

File: Application/resource_manager.py (hunk 1 above)
Comment (ScottCTD) [reply Application/resource_manager.py:178]: u know what, this is a helper function

log is in init_resource**s**

File: Application/data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ScottCTD) [reply Application/data.py:7]: kuo yi

跑回来是因为得用啊 不过logger的initalize还是在main的

logger在哪都能用的

File: Application/resource_manager.py (hunk 1 above)
Comment (Lei-Tin) [reply Application/resource_manager.py:178]: Yes but 他就悄悄的下载了三次然后 Fail 了，不会输出到 Log 的吧

Comment (ScottCTD) [reply Application/resource_manager.py:178]: [Here](https://github.com/UofT-CS-Group/CSC110-Final-Project/pull/31/files#diff-89832c68a9280306bf59dc535cd685ec49bf0281c7ada5e2e91f05d491a463edR155)

如果三次都fail了，就会输出log然后raise一个错

你是想要错一次提示一次？行 我加上

**Review by CharlotteChenyy — APPROVED** [review]
Scott永远滴神!!!

File: Application/resource_manager.py (hunk 1 above)
Comment (Lei-Tin) [reply Application/resource_manager.py:178]: Ya, I saw the part, 我在想要不要错一次提示一次，你要它悄悄的跑完三次也可以

**Pull request merged**